1. Basics/task_4/hw4.php

```diff
@@ -0,0 +1,315 @@
+        $i++;
+    }
+    
+    return array(
+        "method" => trim($methodAndURI[0]),
+        "uri" => trim($methodAndURI[1]),
+        "headers" => $headers,
+        "body" => getStringContentAsPairsArray($body, "&", "="),
+    );
+}
+
+$http = parseTcpStringAsHttpRequest($contents);
+processHttpRequest($http["method"], $http["uri"], $http["headers"], $http["body"]);
+
+/**
+ * Gets string as it input and returns a string which let use PHP_EOL as a delimeter
+ * 
+ * @param $string
+ * 
+ * @return string
+ */
+function getStrToParse($string) : string {
+    $str = preg_replace("/\r/", "", $string);
+    $str = preg_replace("/\n/", PHP_EOL, $str);
+
+    return $str;
+}
+
+/**
+ * Removes items which are empty lines from the array
+ * 
+ * @param array $arr
+ * 
+ * @return array
+ */
+function cleanArray($arr) : array {
+    $resultArr = [];
+    
+    foreach($arr as $item) {
+        if(strlen($item) > 0){
+            $resultArr[]=$item;
+        }
+    }
+
+    return $resultArr;
+}
+
+/**
+ * @param string $string
+ * @param string $arrLastItem
+ * 
+ * @return bool
+ */
+function hasBody($string, $arrLastItem) : bool {
+    return preg_match('/Content-Length:/', $string) || !preg_match('/:/', $arrLastItem);
+}
+
+/**
+ * Analyzes request anr return status code for response
+ * 
+ * @param string    $method
+ * @param string    $uri
+ * @param string    $contentType
+ * @param array     $body
+ * 
+ * @return int
+ */
+function getResponseStatusCode($method, $uri, $contentType, $body) : int {
+    
+    // 400 : Bad Request (invalid method)
+    if($method != "POST") return 400;
+
+    // 400 : Bad Request (invalid "Content-type" header)
+    if($contentType != "application/x-www-form-urlencoded" ) return 400;
+
+    // 400 : Bad Request (invalid body)
+    if(!isRequestBodyValid($body)) return 400;
+
+    // 404 : Not Found (invalid uri)
+    if($uri != "/api/checkLoginAndPassword") return 404;
+    
+    // options: 200, 500, 404
+    return getAccessStatus($body);
+}
+
+/**
+ * Checks if array $body contains "login" and "password" as its keys.
+ * 
+ * @param array $body
+ * 
+ * @return bool 
+ */
+function isRequestBodyValid($body) {
+    return (array_key_exists("login", $body) && array_key_exists("password", $body));
+}
+
+/**
+ * Returns one of 200, 404, 500
+ * 
+ * @param array $body
+ * 
+ * @return int
+ */
+function getAccessStatus($body) : int {
+
+    // 500 : Internal Server Error (file not exist)
+    if(!file_exists(FILE_NAME)) return 500;
+
+    $logAndPassPairsAsArr = getFileContentAsPairsArray(FILE_NAME, PHP_EOL, ":");
+
+    foreach($logAndPassPairsAsArr as $login => $password) {
+        
+        if($body['login'] == $login) {
+            if($body['password'] == $password) {
+                
+                // login found, password matched
+                return 200;
+            }
+
+            // login found, but password DIDN'T match
+            else        
```

Comment:
is `else` there really needed?

Reply:
I think "yes in this option". It prevets to read all the long login-password associative array. Most resources do not allow multiple users with the same names(login), so if the login was found and the password was wrong, this ```else break;``` says: "Stop! No need to search more." and breaks the outer loop too.

The other options: resource allow the scheme "one login - some passwords", logins are sorted and intelligent search uses, logins and passwords store in the hash-map.

Reply:
Concept has been was changed. Now there is an operator ```throw``` instead.